docs/data/common-concepts/custom-components/custom-components.md

```diff
@@ -0,0 +1,244 @@
+# Custom slots and subcomponents
+
+<p class="description">Learn how to override parts of the MUI X components.</p>
+
+## What is a slot?
+
+A slot is a part of a component that can be overridden and / or customized.
+
+## Basic usage
+
+### How to override a slot?
+
+You can override a slot by providing a custom component to the `slots` prop:
+
+{{"demo": "CustomSlot.js"}}
+
+### How to customize a slot?
+
+You can pass props to any slot using the `slotProps` prop:
+
+{{"demo": "CustomSlotProps.js"}}
+
+You can also use both `slots` and `slotProps` on the same component:
+
+{{"demo": "CustomSlotAndSlotProps.js"}}
+
+Most components also support a callback version of `slotProps`.
+This callback receives an object that contains information about the current state of the component,
+that information can vary depending on the slot being used:
+
+{{"demo": "CustomSlotPropsCallback.js"}}
+
+## Correct usage
+
+A slot is a React component; therefore, it should keep the same JavaScript reference between two renders.
+If the JavaScript reference of component changes between two renders, React will remount it.
+You can avoid it by not inlining the component definition in the `slots` prop.
+
+The first two examples below are buggy because the calendar header will remount after each keystroke, leading to a loss of focus.
+
+```jsx
+// ❌ The `calendarHeader` slot is re-defined each time the parent component renders,
+// causing the component to remount.
+function MyApp() {
+  const [name, setName] = React.useState('');
+  return (
+    <DateCalendar
+      slots={{
+        calendarHeader: () => (
+          <input value={name} onChange={(event) => setName(event.target.value)} />
+        ),
+      }}
+    />
+  );
+}
+```
+
+```jsx
+// ❌ The `calendarHeader` slot is re-defined each time `name` is updated,
+// causing the component to remount.
+function MyApp() {
+  const [name, setName] = React.useState('');
+
+  const CustomCalendarHeader = React.useCallback(
+    () => <input value={name} onChange={(event) => setName(event.target.value)} />,
+    [name],
+  );
+
+  return <DateCalendar slots={{ calendarHeader: CustomCalendarHeader }} />;
+}
+```
+
+```jsx
+// ✅ The `calendarHeader` slot is defined only once, it will never remount.
+const CustomCalendarHeader = ({ name, setName }) => (
+  <input value={name} onChange={(event) => setName(event.target.value)} />
+);
+
+function MyApp() {
+  const [name, setName] = React.useState('');
+  return (
+    <DateCalendar
+      slots={{ calendarHeader: CustomCalendarHeader }}
+      slotProps={{ calendarHeader: { name, setName } }}
+    />
+  );
+}
+```
+
+## Usage with TypeScript
+
+### Type custom slots
+
+If you want to ensure type safety on your custom slot component,
+you can declare your component using the `PropsFromSlot` interface:
```

Comment:
How would you access the props expected by a given slot without this util?

Reply:
If the type is not exported—in no clear way.
If it is, by importing the type (i.e. `PickersCalendarHeaderProps` in this case).

Reply:
A solution would indeed be to always export the props of a slot AND to have a consistent nomenclature.
Which can be hard to do since the name of the slot do not give you enough information to know the name of the props interface (I am using the `toolbar` slot of the `DateTimePicker`, is the interface `DateTimePickerToolbarProps`? or `PickersToolbarProps`? or `PickersDateTimePickerProps`?)

The main advantage of this util is that you can use the slot interface of the component you are passing the slot to (and this is trivial to import it) to retrieve the props of the slot.

Reply:
Those are great points. 👍 
I agree with them, just wanted to clarify the direction. 👍 
Such a solution is the next level above DX. 💯

Reply:
Your comment forced me to re-think why I created this util in the 1st place so it's always useful :laughing: